File: src/main/java/io/github/realyusufismail/armourandtoolsmod/ArmourAndToolsMod.java
package io.github.realyusufismail.armourandtoolsmod;

import io.github.realyusufismail.armourandtoolsmod.core.itemgroup.ArmourAndToolsGroup;
import io.github.realyusufismail.armourandtoolsmod.datagen.DataGenerators;
import io.github.realyusufismail.realyusufismailcore.RealYusufIsmailCore;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.Block;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Mod(ArmourAndToolsMod.MOD_ID)
@Mod.EventBusSubscriber(modid = ArmourAndToolsMod.MOD_ID, bus = Mod.EventBusSubscriber.Bus.MOD)
public class ArmourAndToolsMod {
    public static final Logger logger = LogManager.getLogger();
    public static final String MOD_ID = "armourandtoolsmod";

    public static final DeferredRegister<Item> ITEMS =
            DeferredRegister.create(ForgeRegistries.ITEMS, MOD_ID);

    public static final DeferredRegister<Block> BLOCKS =
            DeferredRegister.create(ForgeRegistries.BLOCKS, MOD_ID);

    public ArmourAndToolsMod() {
        IEventBus bus = FMLJavaModLoadingContext.get().getModEventBus();

        ITEMS.register(bus);
        BLOCKS.register(bus);

        // Register the item to a creative tab
        bus.addListener(ArmourAndToolsGroup::registerCreativeTab);
        // Register the data generators
        bus.addListener(DataGenerators::gatherData);

        // Register ourselves for server and other game events we are interested in
        MinecraftForge.EVENT_BUS.register(this);

        logger.info("Loaded Armour and Item Mod");
    }
}

File: src/main/kotlin/io/github/realyusufismail/armourandtoolsmod/ArmourAndToolsMod.kt
/*
 * Copyright 2022 RealYusufIsmail.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */ 
package io.github.realyusufismail.armourandtoolsmod

import io.github.realyusufismail.armourandtoolsmod.client.ClientSetup
import io.github.realyusufismail.armourandtoolsmod.core.init.*
import io.github.realyusufismail.armourandtoolsmod.core.itemgroup.ArmourAndToolsGroup
import io.github.realyusufismail.armourandtoolsmod.datagen.DataGenerators
import java.util.*
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.fml.common.Mod
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import thedarkcolour.kotlinforforge.KotlinModLoadingContext
import thedarkcolour.kotlinforforge.forge.FORGE_BUS

const val MOD_ID = "armourandtoolsmod"

@Mod(MOD_ID)
class ArmourAndToolsMod {
    init {
        val bus = KotlinModLoadingContext.get().getKEventBus()
        ItemInit.ITEMS.register(bus)
        BlockInit.BLOCKS.register(bus)
        MenuTypeInit.MENU.register(bus)
        RecipeSerializerInit.SERIALIZERS.register(bus)
        RecipeTypeInit.RECIPE_TYPES.register(bus)

        // Register the item to a creative tab
        bus.addListener(ArmourAndToolsGroup::registerCreativeTab)
        // Register the data generators
        bus.addListener(DataGenerators::gatherData)
        // adds recipe category
        bus.addListener(RecipeCategoriesInit::registerRecipeBookCategories)
        // client setup listener
        bus.addListener(ClientSetup::clientSetup)
        // register shields
        bus.addListener(ClientSetup::shieldInit)

        // Register ourselves for server and other game events we are interested in
        FORGE_BUS.register(this)
        logger.info("Loaded Armour and Item Mod")
    }

    companion object ArmorAndToolsMod {
        val logger: Logger = LoggerFactory.getLogger(ArmourAndToolsMod::class.java)

        fun getModIdAndName(name: String): ResourceLocation {
            logger.info(
                "Getting mod id and name" +
                    ResourceLocation(MOD_ID, name.lowercase(Locale.getDefault())).toDebugFileName())
            return ResourceLocation(MOD_ID, name.lowercase(Locale.getDefault()))
        }

        fun getPath(path: String): ResourceLocation {
            return ResourceLocation(MOD_ID, path)
        }
    }
}
